### ISSUE TITLE
Add support for `wasm32-unknown-unknown`

### ISSUE THREAD

**Issue by K4rakara** [khvzak/mlua#23]
At the moment, attempting to build for `wasm32-unknown-unknown` causes the following panic: `don't know how to build Lua for wasm32-unknown-unknown`.

I know running Lua in wasm isn't the most efficient thing in the world, but I feel like the functionality should be added.

I'd be happy to look into this, I just want to get the OK before I go and implement it.

**Comment by khvzak** [comment]
The panic is generated by the [`lua-src`](https://github.com/khvzak/lua-src-rs) crate. So it should be a first place for a fix.
Personally, I don't have any objections having lua for the `wasm32` target, but have feeling that this could be quite complicated.

**Comment by lenscas** [comment]
There seem to be 3 main projects to get lua working on wasm
https://github.com/vvanders/wasm_lua
https://github.com/serprex/luwa
https://github.com/ceifa/wasmoon

with https://github.com/ceifa/wasmoon seeming to be the only one that recently had changes

**Comment by makspll** [comment]
I would love to see this implemented!

**Comment by lenscas** [comment]
> I would love to see this implemented! Could somebody summarise the necessary tasks needing to be done for this to run on wassm ?

Step 1: get Lua to compile to WASM or use one of the existing projects for that.

Step 2: find a way to correctly compile said lua VM to wasm together with mlua. Presumably you want them to get compiled in such a way that they share the same WASM memory (So... probably need to end up with 1 wasm file?) to reduce the overhead of FFI.

Step 3: Find some way to get mlua to communicate with the bundled lua. (I am assuming that the normal API's won't work, or that at least some extra plumbing is required.)


Also, there are JS implementations of Lua. Those might be easier to get working  but are probably slower than a version compiled in lua, especially if the wasm version can use the same memory as the rest of the program. https://github.com/fengari-lua/fengari-web is used by the teal playground.



Also, assuming you indeed want to reduce FFI overhead as much as possible and thus make the luavm use the same memory as the rest of the program, then I would imagine that at most only 1 instance of the VM is running. As a result, the functions to create a new Lua instance should probably be disabled and instead replaced with one to get a global instance.

**Comment by makspll** [comment]
Thanks @lenscas!

Hmm would having many lua vm's be **very** bad in terms of efficiency?  ideally the API stays the same between targets, sharing VM's between scripts is very messy and foot-gunny

**Comment by lenscas** [comment]
I would expect that if you want to spawn new VM's that they all have their own memory. Which means that the only way to "share" memory is by copying it over. I'm also not sure if 2 WASM modules can communicate directly with each other. If they can't, you are looking at going through JS. Which means that every value needs to be copied twice.

Lua itself has ways to do sandboxing. Though doing that properly is of course not the easiest.

**Comment by coderedart** [comment]
I will try to add some info.
as @lenscas mentioned, fengari is just a lua vm implemented in JS. wasmoon is a wrapper around lua54 C VM compiled into wasm, and drives it using JS. 

anyway, both of them are useless for our usecase. if you separate out the lua vm from the main rust app as two individual wasm modules, there's nothing for mlua to do. both modules need to share memory and communicate with each other smh, but its more like a lua app and rust app talking to each other and not "embedding" lua into rust app.

I cannot know for sure, but you should be able to embed a lua vm inside rust wasm (no JIt though).  there's two ways:
First, use a rust lua VM like [Hematita](https://github.com/danii/hematita). this should allow us to use `wasm32-unknown-unknown` target + `wasm_bindgen` (which allows rust <-> web JS interactions. used by most tools like `trunk` or eframe). 
BUT apparently, the ABI used is not C ABI compatible. and fixing it would be breaking change for `wasm_bindgen` and its not happening anytime soon. more info https://github.com/rustwasm/wasm-bindgen/pull/2209 . i am not completely sure if hematita will help, gotta ask the developer himself, but the project is being put on hold for the moment. 

Second, use a C lua VM (bindings) and compile it to wasm. like i said, `wasm32-unknown-unknown` toolchain is not available for C code. so, we must use `wasm32-unknown-emscripten` which brings us to the issue https://github.com/rustwasm/team/issues/291 . this seems like the most doable thing at the moment. 

now, my C/C++ (or cmake) skills are absolute garbage, so, dealing with lua54's building process (by copying from wasmoon) is too hard for me. BUT, fortunately, `luau_src` crate used cmake and i was able to successfully do a hello world example using mlua + luau using emscripten target with basically zero effort. 

1. just clone mlua repo.
2. add a basic `main.rs` in the src folder 
```rust
use mlua::Lua;

fn main() {
    println!("inside main");
    for i in 0..5 {
        Lua::new()
            .load(&format!("print('hello lua, times: {i}')"))
            .exec()
            .expect("failed to execute lua");
    }
}

```
3. compile the above with `cargo build --features=luau,vendored --target=wasm32-unknown-emscripten ` to produce a `mlua.js` and `mlua.wasm` files.
4. create a basic html page in crate root folder with `python -m http.server`. don't forget to use the right path for the `mlua.js` script!!!
```html
<script src="./target/wasm32-unknown-emscripten/debug/mlua.js"></script>
<h1>Hello World</h1>

```
5. check console logs to see if there's the `hello lua` output

![image](https://user-images.githubusercontent.com/24411704/180315556-d5a60272-b80d-4728-b977-6ca19650419f.png)

now, technically, lua54 is also doable because wasmoon can do it. but i will leave that task to someone who can understand what's happening in https://github.com/ceifa/wasmoon/blob/main/build.sh . 

just to point out again, this issue is about compilation to `wasm32-unknown-unknown` which is not possible due to ABI issues (which will be fixed at some point according to the linked issue). i don't really understand these things completely, but if you want to know more, there's zulip for wasmtime which has nice helpful people, and there's rust-wasm team channel on officlal rust discord server. 

PS: https://github.com/khvzak/mlua/issues/40 seems to be a duplicate of this.

**Comment by coderedart** [comment]
a lot of crates usually only support `wasm32-unknown-unknown` and seems like winit does the same :crying_cat_face: which makes this useless for embedding in apps / games. 
but emscripten already supports GLFW / SDL to a decent extent https://emscripten.org/docs/porting/multimedia_and_graphics/OpenGL-support.html#opengl-es-extensions , so if rust SDL2 library gets emscripten support https://github.com/Rust-SDL2/rust-sdl2/issues/884 this will be tremendously useful for ~me~ everyone who wants lua scripting on rust web apps  / games.

**Comment by makspll** [comment]
@coderedart fantastic and thorough exploration, thank you!

winit emscripten support seems to be a no-go, and so for engines like bevy I imagine it's an unlikely solution, so it looks like the only way for this to be useful in engines like `bevy` is via this issue
-https://github.com/rustwasm/wasm-bindgen/pull/2209

**Comment by coderedart** [comment]
reply in that thread
> Nope, there hasn't been any progress :( Everyone working in this space has been busy with other things.

i guess this issue will go back to hibernation. i will just keep my eye on hematita.

**Comment by khvzak** [comment]
I plan soon to start porting Lua 5.3 from C to Rust which should bring `wasm32-unknown-unknown` as well as `no_std` (hopefully) support to mlua.
Seems it's the only way (apart from rewriting from scratch).
Why Lua 5.3? Because it's fairly modern, fast and less complex than Lua 5.4 (20k vs 24k LoC).
Also the goal is to keep C API layer backward compatible.

> i will just keep my eye on hematita.
hematita is mostly a POC than a real implementation. It even does has GC support.

**Comment by coderedart** [comment]
I am happy to get anything, but just wanted to know why lua 5.3 instead of 5.1 (which works on luau / luajit)? the whole internet including apps like [mpv ](https://github.com/mpv-player/mpv/issues/9712) seem to stick to 5.1 / 5.2 because of backwards compatibility.

**Comment by lenscas** [comment]
I rather see 5.3 on the Web than 5.1.

There js a compatible library that brings 5.1 and luaJIT to 5.3 compatibility wise. I don't think something similar exists for 5.3 to 5.1.

**Comment by makspll** [comment]
That would be interesting, but also sounds like a big project,

Are there any attempts at bringing lua to pure rust which are somewhat reusable? It's probably worth reaching out to people who have tried this before.

**Comment by coderedart** [comment]
https://github.com/cjneidhart/lua-in-rust

Seems to be a wip lua impl

**Comment by khvzak** [comment]
There are many attempts to implement Lua on pure Rust but unfortunately all of them are not finished and/or ready for production use.
Currently I work on my own implementation of Lua 5.3 and hope next version of mlua will include it. I take a look of what is done to get some ideas.

Unfortunately Lua is not very easy to port/rewrite. There are lots of macros in the source code and tricks.

**Comment by coderedart** [comment]
tracking issue for C ABI and Wasm ABI merge. https://github.com/rust-lang/rust/issues/83788

it seems the work is primarily blocked by clang's multi value return types support in C ABI for wasm.

**Comment by coderedart** [comment]
FWIW, I was able to get both sdl2 and glfw working along with mlua on the emscripten target :) 

a very minimal example https://github.com/coderedart/rust-sdl2-wasm is now live at https://coderedart.github.io/rust-sdl2-wasm/
you should see canvas changing color from black to white over the course of 3 seconds and then reset to black again.
i just get time, and divide the time in seconds by 3. then, i just pass that value to `glClearColor`.

1. go to the above link. 
2. open web developer tools -> console
3. click anywhere on canvas. 
4. the lua code will "print" that value (used for `glClearColor`) in console.

sdl2 basically works without us doing much. 
glfw crate will work too, but will require some tiny patches (mostly feature gating code based on target). 

egui + sdl2 + glow should work on emscripten target. but will need some time to actually try it out. 

I used `Luau` backend as it works perfectly. this will be fine until we get a pure lua implementation of rust.

**Comment by coderedart** [comment]
Finally, It is done. 
egui + sdl2 + emscripten + mlua + Luau + glow
all of them wrapped with a tiny little example inside 50 lines of main.rs
here's a sample:
https://user-images.githubusercontent.com/24411704/216026730-ea65a1d8-a037-4446-b7f2-29f18f446e08.mp4


you can play with it too. live version @ https://coderedart.github.io/rust-sdl2-wasm/
source code @ https://github.com/coderedart/rust-sdl2-wasm .  much simpler than you would expect :D

the next step is to wait for mlua 0.9 when we can actually script egui with lua **live* inside a browser.

**Comment by tari** [comment]
https://github.com/khvzak/lua-src-rs/pull/7 make `lua-src` build for emscripten targets, and using the changes in https://github.com/khvzak/mlua/pull/337 I've been able to build an application embedding Lua 5.4 for `wasm32-unknown-emscripten`. Luau worked without any changes as well.

**Comment by Andrewp2** [comment]
Does https://github.com/rust-lang/rust/pull/117919 `-Zwasm-c-abi` matter at all for this issue?

**Comment by elliotsayes** [comment]
@Andrewp2 any luck with this? Seems like it should work but I can't work out how to integrate it

**Comment by elliotsayes** [comment]
@Andrewp2 did you happen to get this working? Seems like it should be possible, but not sure how

**Comment by TannerRogalsky** [comment]
I would only expect the WASM C ABI to be relevant for bindings. There are two issues, I'd guess.

1. Compiling Lua with a wasm32-unknown-unknown standard library. Easiest way is probably with wasi.
2. Compiling Lua with setjmp/longjmp support. I don't know what the state of the wasi exceptions work is but this would probably need it.

**Comment by LqdBcnAtWork** [comment]
> 2. Compiling Lua with setjmp/longjmp support. I don't know what the state of the wasi exceptions work is but this would probably need it.

Looks to be Phase 4, with Chrome/Node/Deno needing a flag, Firefox and Safari have full support. Pretty much no support elsewhere.

https://webassembly.org/features/

I would love a `wasi32-unknown-unknown` target (especially with luau). I'd use it as a user facing scripting option for my web/electron app that uses rust compiled to wasm for data parsing.

I'm just hopeless with C/C++

**Comment by Bobbyjoness** [comment]
> Looks to be Phase 4, with Chrome/Node/Deno needing a flag, Firefox and Safari have full support. Pretty much no support elsewhere.

Does this mean that technically that mlua can be compiled using wasi or do we need to wait for rust side of things to do something to enable support?

**Comment by cheesycod** [comment]
@khvzak  any update on getting Luau at least to work with wasm32-unknown-unknown as this would be useful to something I'm working on right now

**Comment by anthonydito** [comment]
I am very interested in this working with `wasm32-unknown-unknown` for [my project](https://www.brushcue.com/). If there is anything I can do to help or any pointers to work on this, please let me know!

**Comment by Auxnon** [comment]
> I am very interested in this working with `wasm32-unknown-unknown` for [my project](https://www.brushcue.com/). If there is anything I can do to help or any pointers to work on this, please let me know!

Can i plug kyren's https://github.com/kyren/piccolo ? She was the original creator of rlua iirc and it's a completely pure rust solution, compiles to wasm32-unknown-unknown, and relies on a garbage collection arena with it's own lifetime and seems to cover everything PUC Lua offers.

( I also made a shoddy Lua interpreter that I hope to add luau-style type casting to eventually but it's order of magnitudes worse then piccolo https://github.com/auxnon/silt-lua  I wanted it to be drop-in replacement for mlua for wasm builds but honestly just use piccolo lol)

**Comment by speedy-lex** [comment]
I made a custom version of lua53-sys that compiles a patched lua53 with a custom libc and libcppthrow. It works and compiles for wasm32-unknown-unknown. https://github.com/speedy-lex/lua53-sys

**Comment by LqdBcnAtWork** [comment]
An update on wasm exceptions, it's up to Phase 5, "The Feature is Standardized".

As of writing, of the 14 runtimes monitored on webassemby.org, 9 have full support, 2 have experimental, and 3 have no support.

So I don't know what else is needed for it, but C++ exceptions via WASM EH should only need the right flags set now.  
https://emscripten.org/docs/tools_reference/settings_reference.html#support-longjmp  
https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fwasm-exceptions  

I believe we just need shims to supplement `mlua`s needs from the std lib. On either the Rust or C++ side. Which, unless I'm wrong, should mainly be some File IO related functions.

**Comment by CppCXY** [comment]
Perhaps https://github.com/CppCXY/lua-rs is another option, as it can be compiled to wasm-unknown-unknown.

<img width="1885" height="902" alt="Image" src="https://github.com/user-attachments/assets/05601f4c-b8ab-4325-b2dd-eccf8931a71d" />

**Comment by lenscas** [comment]
> Perhaps https://github.com/CppCXY/lua-rs is another option, as it can be compiled to wasm-unknown-unknown.
> 
> <img alt="Image" width="1885" height="902" src="https://private-user-images.githubusercontent.com/40318218/582505881-05601f4c-b8ab-4325-b2dd-eccf8931a71d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY5Mjk3NjUsIm5iZiI6MTc3NjkyOTQ2NSwicGF0aCI6Ii80MDMxODIxOC81ODI1MDU4ODEtMDU2MDFmNGMtYjhhYi00MzI1LWIyZGQtZWNjZjg5MzFhNzFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDIzVDA3MzEwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxOTM1OWQ5NmI1Y2U4Y2NjMGYzNGMwMWIyZjU0MDcyYWVhZDVjOTgxM2U3MmM3Y2M0M2FhNjc1OTQwZTg1Y2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.9RFlrt18Y2Qpb-bweFdcPN5Pa96BOBboe_eTXHQbGvg">

But why go for a vibe coded version of Lua that, unless I am mistaken adds nothing new to the table? As stated above, the missing pieces on the Wasm side should be fixed now and the ABI problems that rust has are in the process of being fixed or already fixed.

There are fewer and fewer blockers basically everyday and it might already be possible to get it all working if someone is willing to spend the time.

If a rust based implementation is needed then why not something like https://github.com/lumen-oss/ottavino ?

It is already used by projects (both the project it forked and the fork I linked) and brings new functionality to the table aside from the usual "it is written in rust"

**Comment by CppCXY** [comment]
> But why go for a vibe coded version of Lua that, unless I am mistaken adds nothing new to the table? As stated above, the missing pieces on the Wasm side should be fixed now and the ABI problems that rust has are in the process of being fixed or already fixed.
> 
> There are fewer and fewer blockers basically everyday and it might already be possible to get it all working if someone is willing to spend the time.
> 
> If a rust based implementation is needed then why not something like https://github.com/lumen-oss/ottavino ?
> 
> It is already used by projects (both the project it forked and the fork I linked) and brings new functionality to the table aside from the usual "it is written in rust"

From a performance perspective, this is currently the fastest, and it passes the vast majority of the official Lua test suite. From an implementation perspective, it is the most complete. Even the project you mentioned, https://github.com/lumen-oss/ottavino, has its string algorithms ported from lua-rs.
<img width="661" height="233" alt="Image" src="https://github.com/user-attachments/assets/02f46cb9-bbcd-4a96-a2f6-27991c2c8bdd" />

**Comment by ChedRed** [comment]
What's the current status on this? Is there any way I can try to build mlua for wasm32-unknown-unknown right now?

**Comment by lenscas** [comment]
> What's the current status on this? Is there any way I can try to build mlua for wasm32-unknown-unknown right now?

From my understanding, everything on the rust and Wasm specification side is now pretty much resolved. So the next step is that someone takes (nightly) rust and tries it out. Then seeing what they run into, if it can be solved by passing specific arguments to the compiler or if something more involved is needed. And of course document it.

**Comment by ChedRed** [comment]
I tried building with `cargo +nightly build --target wasm32-unknown-unknown`, but it failed at mlua-sys, which does not know how to build for wasm32-unknown-unknown. I only added mlua to the Cargo.toml, and I tried with the latest version and directly from the github repo. I'm not too sure where to go from here, especially since im unfamiliar with testing crates. Do I also need to include a specific version of mlua-sys or lua itself?

**Comment by izissise** [comment]
@speedy-lex [version](https://github.com/mlua-rs/mlua/issues/23#issuecomment-3826535546) compiles for wasm32-unknown-unknown you can easily integrate it with mlua by changing some lines in Cargo.toml

**Comment by khvzak** [comment]
I'm working on `wasm32-unknown-unknown` support. I have demo with transpiled/ported Lua 5.5 to Rust. The only problem it requires `c_variadic` feature that should be stable in Rust 1.99 (https://github.com/rust-lang/rust/issues/44930)

**Comment by speedy-lex** [comment]
> > What's the current status on this? Is there any way I can try to build mlua for wasm32-unknown-unknown right now?
> 
> From my understanding, everything on the rust and Wasm specification side is now pretty much resolved. So the next step is that someone takes (nightly) rust and tries it out. Then seeing what they run into, if it can be solved by passing specific arguments to the compiler or if something more involved is needed. And of course document 

I already have tested and used it in wasm32-unknown-unknown but it was almost a year ago so a lot of stuff can probably be removed.
See: https://github.com/speedy-lex/lua53-sys
